Make splashScreenView of FlutterViewController nullable (#7385)
flutter/website: src/release/breaking-changes/ios-flutterviewcontroller-splashscreenview-nullable.md

---
title: iOS FlutterViewController splashScreenView make nullable
description: FlutterViewController splashScreenView changed from nonnull to nullable.
---

## Summary

We changed property `splashScreenView` of `FlutterViewController` from `nonnull` to `nullable`.

Old declaration of `splashScreenView`:

```objective-c
@property(strong, nonatomic) UIView* splashScreenView;
```

New declaration of `splashScreenView`:

```objective-c
@property(strong, nonatomic, nullable) UIView* splashScreenView;
```

## Context

This allows iOS add-to-app users writen in `Swift` to remove the splash screen view.

However, the previous declaration would build error when set `nil` value using `Swift`, only supported if using `Objective-C`.

## Description of change

> Before migrated, it would be a compilation error if the calling code is in Swift and gets the splash screen and stores in a nonnull UIView variable.
> 
> ```
> error build: Value of optional type 'UIView?' must be unwrapped to a value of type 'UIView'
> ```

Reviewer: ````suggestion
While it was possible in Objective-C to work around the incorrect `nonnull` annotation by setting
`splashScreenView` to a `nil` `UIView`, in Swift this caused a compilation error:
```
error build: Value of optional type 'UIView?' must be unwrapped to a value of type 'UIView'
```

[PR #34743][] updates the property attribute to `nullable`. It can return `nil` and can be set to `nil` to remove the view in both Objective-C and Swift.
````

> After [PR #34743][], you need to update code to instead store in a UIView?. Instead of declare variable `splashScreenView` as type `UIView`:

Reviewer 2: ```suggestion
After [PR #34743][], the type of `splashScreenView` becomes `UIView?`:
```

> 
> After [PR #34743][], you need to update code to instead store in a UIView?. Instead of declare variable `splashScreenView` as type `UIView`:
> 
> ```swift
> @UIApplicationMain
> @objc class AppDelegate: FlutterAppDelegate {
>   var splashScreenView = UIView()
>   override func application(
>     _ application: UIApplication,
>     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
>   ) -> Bool {
>     var flutterEngine = FlutterEngine(name: "my flutter engine")
>     let flutterViewController = FlutterViewController(engine: flutterEngine, nibName: nil, bundle: nil)
>     splashScreenView = flutterViewController.splashScreenView // compilation error: Value of optional type 'UIView?' must be unwrapped to a value of type 'UIView'
> ```
> 
> instead use `UIView?`:
> 
> ```swift
> @UIApplicationMain
> @objc class AppDelegate: FlutterAppDelegate {
>   var splashScreenView : UIView? = UIView()
>   override func application(
>     _ application: UIApplication,
>     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
>   ) -> Bool {
>     var flutterEngine = FlutterEngine(name: "my flutter engine")
>     let flutterViewController = FlutterViewController(engine: flutterEngine, nibName: nil, bundle: nil)
>     let splashScreenView = flutterViewController.splashScreenView // compiles successfully
>     if let splashScreenView = splashScreenView {
>       
>     }
> ```
> 

Reviewer: I think this part can be moved to "Migration guide".
```suggestion
```

## Migration guide

> Code before migration:

Reviewer: ```suggestion
If `splashScreenView` is stored in a `UIView` variable in Swift, update to an optional type `UIView?`.
 
Code before migration:
```

```swift
  var splashScreenView = UIView()
  var flutterEngine = FlutterEngine(name: "my flutter engine")
  let flutterViewController = FlutterViewController(engine: flutterEngine, nibName: nil, bundle: nil)
  splashScreenView = flutterViewController.splashScreenView // compilation error: Value of optional type 'UIView?' must be unwrapped to a value of type 'UIView'
```

Code after migration:

```swift
  var splashScreenView : UIView? = UIView()
  var flutterEngine = FlutterEngine(name: "my flutter engine")
  let flutterViewController = FlutterViewController(engine: flutterEngine, nibName: nil, bundle: nil)
  let splashScreenView = flutterViewController.splashScreenView // compiles successfully
  if let splashScreenView = splashScreenView {
  }
```

## Timeline

> In stable release: not yet

Reviewer: ```suggestion
In stable release: TBD
```
